lib/puppet/resource.rb

```diff
@@ -433,12 +435,12 @@ def lookup_with_databinding(name, scope)
   private :lookup_with_databinding
 
   def lookup_in_environment(name, scope)
-    Puppet::DataProviders.lookup_in_environment(name, scope)
+    catch(:no_such_key) { Puppet::DataProviders.lookup_in_environment(name, scope, nil) }
```

Comment:
The implementation will never use the returned value in case an exception is thrown. I thought about having a special marker object but decided not to. This way, neither the caller or an implementer needs to worry about such a value.